apps/server/src/modules/auth/next-auth.controller.ts

```diff
@@ -177,10 +177,17 @@ export class NextAuthController {
       this.metrics.authFailCounter(1, {
         reason: 'no_early_access_permission',
       });
-      if (
-        !req.headers?.referer ||
-        req.headers.referer.startsWith('https://accounts.google.com')
-      ) {
+      const checkReferer = () => {
```

Comment:
Maybe you can define `checkReferer` top of the file or inside the `NextAuthController` class as a private method?

```
const checkReferer = (referer?: string) => {
    if (!referer) return false
    // other logic
}
```

Reply:
I have considered that before.
The thing is this `checkReferer` is not general enough (or robust enough) to be moved to the class level nor module level. I don't want other people to misuse this function.
As we already have
```js
if (!req.headers?.referer || checkReferer()) {
```
it checks the referer whether it's undefined. If undefined, `checkReferer` will be short circuit and not be executed.
Hence, I put
```
new URL(req.headers.referer ?? '')
```
here is just to make compiler happy.

As the function itself depends on the context, I would prefer not moving it outside.

The function you have written is reansonable and intuitive to be put outside.

However, if I would like to extract the original method outside will be

```js
const checkReferer = (referer?: string) => {
 -   if (!referer) return false
 +   if (!referer) return true // to replace !req.headers?.referer
    // other logic
}
```

That will make `checkReferer` function weird. Would you still insist on this?

Reply:
> As the function itself depends on the context, I would prefer not moving it outside.

Okay, considering that the "early access" may not last long and currently only Google is supported, it is acceptable to place `checkReferer` inside the function.

> That will make checkReferer function weird.

If `checkReferer` will reverse the logic, consider reversing `if {} else {}` as well and add some brief comments? or simply consider a more appropriate name, perhaps starting with `should`?

Reply:
I come up with a better solution.
Since we have
```js
req.headers.referer.startwith('https://accounts.google.com')
```
before.

We can just implement a function called `checkRefererOrigin` to replace this in place.

That is
```js
checkRefererOrigin(req.headers.referer, 'https://accounts.google.com')
```

Unfortunately, I have my computer broken recently and not able to modify it. Would you like to open a new pull request to fix this?
Once you open that, I will close this pull request. Thanks a lot.

Reply:
> Would you like to open a new pull request to fix this?

Sorry for the late reply, but I can't help you. I have my own things to deal with.